### PR TITLE
use Graphback commands for "generate", "serve" and "init" commands

### PR DESCRIPTION
**Pull request by machi1990**
/cc @craicoverflow @wtrocki 

Fixes https://github.com/aerogear/graphback/issues/1736

## What was done
 * Problems solved:
   * Te-stability and easy maintainability of the changes. Each command is individual tested on Graphback side and with each release of the commands on Graphback side, this will be automatically picked up by `renovate` on graphql-cli side and be bumped automatically without requiring major work.  
   * Templates that are independent of the version of CLI (we lock version of the templates)
   * Release cadence (we can release now CLI more frequently and we will include it in our testing schedule). 
   * GraphQL-CLI becomes default part of our ecosystem and default way to consume multiple projects related to graphql crud
   * Now the `serve` commands, serves a GraphQL server which is backed by an in memory database
* What is the impact of the changes
   * This is still nonmajor change etc.
 * Benefits of the changes. For example:
   * We brought customization and custom resolvers
   * We enabled cogen typings bringing more development use cases
   * We added more data sources: MongoDB, Postgress, DataSync
   * Better documentation for the serve command

### CODE REVIEW

**Comment by wtrocki** [comment]
We need to add small changes to init as well

**Comment by machi1990** [comment]
> We need to add small changes to init as well

@craicoverflow @wtrocki I've made small adjustments in the init command too. This should be ready for a full review.

**Comment by wtrocki** [comment]
@machi1990 We need to have a proper description of the changes in the PR before we merge and some instruction on how to 

- What was done
- What is the impact of the changes
- Benefits of the changes. For example:
   - We brought customization and custom resolvers 
   - We enabled cogen typings bringing more development use cases
   - We added more data sources: MongoDB, Postgress, DataSync
   - Better documentation 

What we also need is to update website docs to point out to required resources (model creation etc.)

**Comment by wtrocki** [comment]
CC @ardatan

**Comment by machi1990** [comment]
> @machi1990 We need to have a proper description of the changes in the PR before we merge and some instruction on how to verify this changes.
> 
> * What was done
> * Problems with solved:
>   
>   * Testability of the changes
>   * Templates that are independent of the version of CLI (we lock version of the templates)
>   * Release cadence (we can release now CLI more frequently and we will include it in our testing schedule)
>   * GraphQL-CLI becomes default part of our ecosystem and default way to consume multiple projects related to graphql crud
> * What is the impact of the changes
>   
>   * This is still nonmajor change etc.
> * Benefits of the changes. For example:
>   
>   * We brought customization and custom resolvers
>   * We enabled cogen typings bringing more development use cases
>   * We added more data sources: MongoDB, Postgress, DataSync
>   * Better documentation
> 

This is well put, thanks. Do you want me to copy it over the PR description or it's visible enough where it is?

> What we also need is to update website docs to point out to required resources (model creation etc.)

I am only seeing the `docs` folder which contains two file. I am sure where to add docs for changes brought by this PR. Any pointer are welcomed.

**Comment by wtrocki** [comment]
> I am only seeing the docs folder which contains two file. I am sure where to add docs for changes brought by this PR. Any pointer are welcomed.

I have merged website PR. Please check master.

> This is well put, thanks. Do you want me to copy it over the PR description or it's visible enough where it is?

Copy what I have put as example but please extend it when needed. I wanted to put some guidance

**Comment by machi1990** [comment]
> > I am only seeing the docs folder which contains two file. I am sure where to add docs for changes brought by this PR. Any pointer are welcomed.
> 
> I have merged website PR. Please check master.
> 

Nice. I guess this is what was missing. I'll jump on it later.

**Comment by craicoverflow** [comment]
```patch
"workspaces": {
    "packages": [
      "packages/*",
      "packages/commands/*",
-     "integration"
+     "integration/*"
    ]
},
```

It's not in scope here, but the root package.json will require the integration subfolders to be included in workspaces, as I cannot run `integration/test-project` without it (unless I install node_modules locally)

**Comment by machi1990** [comment]
> ```diff
> ```diff
> "workspaces": {
>     "packages": [
>       "packages/*",
>       "packages/commands/*",
> -     "integration"
> +     "integration/*"
>     ]
> },
> ```
> 
> 
> It's not in scope here, but the root package.json will require the integration subfolders to be included in workspaces, as I cannot run integration/test-project without it (unless I install node_modules locally)
> ```

Sure, I’ll add this change. As a side note, the ‘test-project’ is just a fake nodejs project added for testing the workflow. It is not meant to bs run as a stand alone project. This is after the deletion of ‘templates’ folder.

File: integration/test-project/.graphqlrc.yml [hunk 1]
```diff
@@ -1,20 +1,8 @@
...
-          skipDocumentsValidation: true
-        plugins:
-          - add: '/* tslint:disable */'
-          - typescript
-          - typescript-operations
-          - typescript-react-apollo
-      ./server/src/generated-types.ts:
+      ./src/generated-types.ts:
         config:
           contextType: '@graphback/runtime#GraphbackRuntimeContext'
```
Comment (craicoverflow) [inline integration/test-project/.graphqlrc.yml:7]: `@graphback/runtime#GraphbackRuntimeContext` package interface does not exist any more. `GraphbackContext` now sits in `@graphback/core`, so maybe:

```suggestion
          contextType: '@graphback/core#GraphbackContext'
```

File: integration/test-project/package.json [hunk 2]
```diff
@@ -1,16 +1,32 @@
...
-    "start:server": "cd server/ && yarn start",
-    "start:client": "cd client/ && yarn start"
+    "develop": "ts-node-dev src/index.ts",
+    "start": "ts-node src/index.ts"
+  },
+  "dependencies": {
+    "graphback": "0.15.1",
+    "@graphback/runtime-knex": "0.15.1",
+    "@graphback/codegen-schema": "0.15.1",
+    "graphql-migrations": "0.15.1",
```
Comment (craicoverflow) [inline integration/test-project/package.json:13]: Should these be bumped to 0.16.1?

File: integration/test-project/package.json [hunk 3]
```diff
@@ -1,16 +1,32 @@
...
+    "apollo-server-express": "2.17.0",
+    "cors": "2.8.5",
+    "express": "4.17.1",
+    "graphql": "15.3.0",
+    "graphql-config": "3.0.3",
+    "graphql-tag": "2.11.0",
+    "knex": "0.21.5",
+    "pg": "8.3.3",
+    "ts-node": "8.10.2",
+    "ts-node-dev": "1.0.0-pre.61",
```
Comment (craicoverflow) [inline integration/test-project/package.json:27]: Why is this needed along with ts-node?

Comment (machi1990) [reply integration/test-project/package.json:27]: Not needed.

File: integration/test-project/package.json (hunk 2 above)
Comment (craicoverflow) [reply integration/test-project/package.json:13]: Ignore, that will be a follow up :)

**Comment by craicoverflow** [comment]
When I run `init`, I am prompted to choose What type of project I want. This question is disregarded later and I am asked to choose one from all the Graphback templates. Maybe the first question should be removed?

```bash
> yarn graphql init
yarn run v1.22.4
$ /home/ephelan/code/Urigo/graphql-cli/node_modules/.bin/graphql init
? What is the type of the project? Backend only
```

**Comment by machi1990** [comment]
> When I run `init`, I am prompted to choose What type of project I want. This question is disregarded later and I am asked to choose one from all the Graphback templates. Maybe the first question should be removed?
> 
> ```shell
> > yarn graphql init
> yarn run v1.22.4
> $ /home/ephelan/code/Urigo/graphql-cli/node_modules/.bin/graphql init
> ? What is the type of the project? Backend only
> ```

The project type question is used later on to initialise codegen related bits and it won't have much influence on the command brought by graphback. It will be interesting to pass this variable later onto the create-graphback handler to filter projects accordingly but I propose that to be on a followup PR.

**Comment by machi1990** [comment]
> > When I run `init`, I am prompted to choose What type of project I want. This question is disregarded later and I am asked to choose one from all the Graphback templates. Maybe the first question should be removed?
> > ```shell
> > > yarn graphql init
> > yarn run v1.22.4
> > $ /home/ephelan/code/Urigo/graphql-cli/node_modules/.bin/graphql init
> > ? What is the type of the project? Backend only
> > ```
> 
> The project type question is used later on to initialise codegen related bits and it won't have much influence on the command brought by graphback. It will be interesting to pass this variable later onto the create-graphback handler to filter projects accordingly but I propose that to be on a followup PR.

I've opened thttps://github.com/aerogear/graphback/issues/1981 to track this

File: website/docs/command-serve.md [hunk 4]
```diff
@@ -37,18 +39,108 @@ import TabItem from '@theme/TabItem';
...
+
+For the `Note` model defined above, this would look like: 
+
+```graphql
+""" 
+@model
+@datasync 
+"""
+type Note {
+  id: ID!
```
Comment (craicoverflow) [inline website/docs/command-serve.md:101]: Data sync will fail unless you use `_id: GraphbackObjectID` since this is MongoDB.

```suggestion
  _id: GraphbackObjectID!
```

File: website/docs/command-serve.md [hunk 5]
```diff
@@ -37,18 +39,108 @@ import TabItem from '@theme/TabItem';
 
 ### Usage
 
+The bare minimum you need is a GraphQL file with your data models. Create a file called `Note.graphql` and add the following:
+
+```graphql
+""" @model """
+type Note {
+  id: ID!
```
Comment (craicoverflow) [inline website/docs/command-serve.md:47]: Since GraphQL Serve uses MongoDB, it should be `_id: GraphbackObjectID`

**Comment by wtrocki** [comment]
I think we discussed that init generally should be more specific to project command. with that in mind me can first ask questions for graphback and then use codegen.

Something like:
- do you want to start with template (graphback,)
- do you want to add cli to existing project

And later codegen, diff questions

File: website/docs/command-serve.md (hunk 4 above)
Comment (machi1990) [reply website/docs/command-serve.md:101]: Good catch!

**Comment by wtrocki** [comment]
Reviewing now

**Review by wtrocki — APPROVED** [review]

**Comment by wtrocki** [comment]
Once merged to master we can use released version (canary tag) for full scope testing
